stan/math/prim/mat/fun/gp_dot_prod_cov.hpp

```diff
@@ -93,10 +102,12 @@ gp_dot_prod_cov(const std::vector<T_x> &x, const T_sigma &sigma) {
  * @throw std::domain_error if sigma < 0, nan, inf or
  *   x is nan or infinite
  */
-template <typename T_sigma>
-inline typename Eigen::Matrix<typename stan::return_type<double, T_sigma>::type,
-                              Eigen::Dynamic, Eigen::Dynamic>
-gp_dot_prod_cov(const std::vector<double> &x, const T_sigma &sigma) {
+template <typename T_x, typename T_sigma>
+inline typename boost::enable_if_c<
+    is_constant<T_x>::value,
```

Comment:
Change the function signature so it returns a matrix. Like:

```
gp_dot_prod_cov(const std::vector<Eigen::Matrix<T_x, Eigen::Dynamic, 1>> &x, const T_sigma &sigma)
```

Let's not worry about ```Eigen::Matrix<T_x, 1, Eigen::Dynamic>``` types